### ISSUE TITLE
<crtdefs.h>: Header is missing from this repo

### ISSUE THREAD

**Issue by BillyONeal** [microsoft/STL#613]
**Describe the bug**
In https://github.com/microsoft/STL/pull/593 @AlexGuteniev indicated that our "__dllimport" macros are coming from `crtdefs.h` which is not presently in the repo; that is part of "vcruntime".

Potential solutions:
* We could do all new satellite DLLs with a .def file and not use any import/export macro.
* We could do all new satellite DLLs with a different header than `crtdefs.h` that lives in this repo.
* We could move vcruntime into github (we already made our directory structure allow for that).

**Comment by AlexGuteniev** [comment]
> * We could move vcruntime into github (we already made our directory structure allow for that).

That sounds interesting as it might help in other places.

I've asked why is `_YIELD_PROCESSOR()` defined as `yield` instruction on ARM:
https://github.com/microsoft/STL/blob/285187b7b24be4bebfa271fbd0b1eacd995d18f4/stl/inc/xatomic.h#L40
but not as `pause` on x86 / x86-64:
https://github.com/microsoft/STL/blob/285187b7b24be4bebfa271fbd0b1eacd995d18f4/stl/inc/xatomic.h#L31

Assuming there's no particular reason, and putting `pause` would be useful, I'll need `_mm_pause` in `<intrin0.h>`.  But again, `<intrin0.h>` is not here on Github.

**Comment by BillyONeal** [comment]
The answer I've heard from folks that supposedly know better is "never `_mm_pause` ever" -- that macro was defined long before my time but it was defined as empty for x86/x64 and I don't think we would want to change that without extremely strong profiling evidence that it was a good idea.

**Comment by AlexGuteniev** [comment]
Ok, then `YieldProcessor()` in SDK headers is probably wrong.
Anyway, I guess I need to seek answer in Intel / AMD manuals.

**Comment by AlexGuteniev** [comment]
(But hey, Intel TBB uses `_mm_pause`! They are supposed to know the best)

**Comment by AlexGuteniev** [comment]
Even `WaitOnAddress` itself uses `pause` for its internal spinning (I don't have access to Windows source, but can step into using a debugger).

I'm not insisting `_YIELD_PROCESSOR()` should be redefined, but I think the full rule of _mm_pause usage is not as simple as "never `_mm_pause` ever"

**Comment by jonwil** [comment]
It would definitely be nice to see the source code to the VC Runtime stuff (vcruntime140.dll, vcruntime.lib, msvcrt.lib and the like) made open source.

I assume crtdefs.h is needed by things other than the STL and moving it into the STL repo isn't an option?

**Comment by StephanTLavavej** [comment]
Yeah, that would be a layering issue. We're looking into open-sourcing almost all of vcruntime/vcstartup but there are many gears that need to turn before that can happen.

**Comment by StephanTLavavej** [comment]
Variant of "We could do all new satellite DLLs with a different header than crtdefs.h that lives in this repo." - we could simply move the relevant macros into `yvals.h` instead of introducing a whole header. (After solving logistics issues related to the move, since MSVC has to ship Previews before we can consume its headers.)

**Comment by AlexGuteniev** [comment]
I think I have fourth idea, which looks better.

Can do this, like .def without .def, in implementation .cpp:
```
#ifdef _BUILDING_SATELLITE_ATOMIC_WAIT

#pragma comment(linker, "/export:__std_atomic_set_api_level")
#pragma comment(linker, "/export:__std_atomic_get_spin_count")
#pragma comment(linker, "/export:__std_atomic_wait_get_deadline")
#pragma comment(linker, "/export:__std_atomic_wait_direct")
...
#endif // _BUILDING_SATELLITE_ATOMIC_WAIT
```

Then don't need to decorate names anyhow, but still using preprocessor condition passed from make file

The best part of it is that I can _conditionally_ use forwarders (#763):

```
#ifdef _ATOMIC_WAIT_STATICALLY_AVAILABLE_TO_IMPL

#pragma comment(linker, "/export:__std_atomic_notify_one_direct=api-ms-win-core-synch-l1-2-0.WakeByAddressSingle")
#pragma comment(linker, "/export:__std_atomic_notify_all_direct=api-ms-win-core-synch-l1-2-0.WakeByAddressAll")

#else // ^^^ statically available / not statically available vvv

#pragma comment(linker, "/export:__std_atomic_notify_one_direct")
#pragma comment(linker, "/export:__std_atomic_notify_all_direct")

#endif // _ATOMIC_WAIT_STATICALLY_AVAILABLE_TO_IMPL


````

**Comment by AlexGuteniev** [comment]
Syntax with `#pragma comment(linker, "/export...")` proved to be not a good idea on x86, when there are name decorations involved dependent on calling conventions. `.def` file somehow bypasses the decoration problem, and `#pragma comment` somehow fails.
Manually specifying decorated name, which would work for `#pragma comment`,  looks like a bad option from maintenance viewpoint, and just looks ugly.
Still `#pragma comment` works well for frowarders -- if there would be a decision to use them.

**Comment by AlexGuteniev** [comment]
I assume this is now `wontfix`, and we'll live with this limitation

**Comment by StephanTLavavej** [comment]
Yeah, unfortunately. Closing.
